File: vital-players/src/main/java/me/vitalframework/players/VitalPlayerListener.java
package me.vitalframework.players;

import jakarta.annotation.PostConstruct;
import lombok.Getter;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;
import me.vitalframework.VitalRepository;
import net.md_5.bungee.api.ProxyServer;
import net.md_5.bungee.api.connection.ProxiedPlayer;
import net.md_5.bungee.api.event.PlayerDisconnectEvent;
import net.md_5.bungee.api.event.PostLoginEvent;
import net.md_5.bungee.api.plugin.Plugin;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.event.EventPriority;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.Optional;
import java.util.UUID;

/**
 * A listener class that manages VitalPlayer instances when players join and leave the server.
 *
 * @author xRa1ny
 */
@RequiredArgsConstructor
@Getter
public abstract class VitalPlayerListener<P, T, VP extends VitalPlayer<?>, PR extends VitalRepository<VP>> {
    @NonNull
    private final P plugin;

    @NonNull
    private final PR playerRepository;

    public void handlePlayerJoin(@NonNull UUID uniqueId, @NonNull T player) {
        // Retrieve the VitalPlayer associated with the joining player, if it exists.
        final var optionalVitalPlayer = Optional.ofNullable(playerRepository.getComponentByUniqueId(uniqueId));

        if (optionalVitalPlayer.isEmpty()) {
            // Create a new VitalPlayer for the joining player.
            try {
                final var vitalPlayer = vitalPlayerType().getDeclaredConstructor(playerType()).newInstance(player);

                // Register the VitalPlayer with VitalUserManagement.
                playerRepository.registerComponent(vitalPlayer);
            } catch (Exception e) {
                e.printStackTrace();
                throw new RuntimeException("error while creating vital player instance %s for %s"
                        .formatted(vitalPlayerType().getSimpleName(), uniqueId));
            }
        }
    }

    public void handlePlayerQuit(@NonNull UUID uniqueId) {
        // Retrieve the VitalPlayer associated with the leaving player.
        final var optionalVitalPlayer = Optional.ofNullable(playerRepository.getComponentByUniqueId(uniqueId));

        if (optionalVitalPlayer.isEmpty()) {
            return;
        }

        final var vitalPlayer = optionalVitalPlayer.get();

        // Unregister the VitalPlayer from VitalUserManagement.
        playerRepository.unregisterComponent(vitalPlayer);
    }

    /**
     * Defines the type this {@link VitalPlayerListener} manages.
     *
     * @return The type of the managed {@link VitalPlayer}.
     */
    @NonNull
    protected abstract Class<VP> vitalPlayerType();

    /**
     * Defines the player type every VitalPlayer extending class takes as an argument for construction.
     *
     * @return The player type for injection.
     */
    @NonNull
    protected abstract Class<T> playerType();

    public static abstract class Spigot<VP extends VitalPlayer.Spigot, PM extends VitalRepository<VP>> extends VitalPlayerListener<JavaPlugin, Player, VP, PM> implements org.bukkit.event.Listener {
        public Spigot(@NonNull JavaPlugin plugin, @NonNull PM playerRepository) {
            super(plugin, playerRepository);
        }

        @PostConstruct
        public void init() {
            // disconnect any connected player...
            for (var player : Bukkit.getOnlinePlayers()) {
                player.kickPlayer("");
            }

            getPlugin().getServer().getPluginManager().registerEvents(this, getPlugin());
        }

        @Override
        protected @NonNull Class<Player> playerType() {
            return Player.class;
        }

        // should always be executed first.
        @org.bukkit.event.EventHandler(priority = EventPriority.LOW)
        public final void onPlayerJoin(@NonNull PlayerJoinEvent e) {
            handlePlayerJoin(e.getPlayer().getUniqueId(), e.getPlayer());
        }

        // should always be executed last.
        @org.bukkit.event.EventHandler(priority = EventPriority.HIGH)
        public final void onPlayerQuit(@NonNull PlayerQuitEvent e) {
            handlePlayerQuit(e.getPlayer().getUniqueId());
        }
    }

    public static abstract class Bungeecord<VP extends VitalPlayer.Bungeecord, PM extends VitalRepository<VP>> extends VitalPlayerListener<Plugin, ProxiedPlayer, VP, PM> implements net.md_5.bungee.api.plugin.Listener {
        public Bungeecord(@NonNull Plugin plugin, @NonNull PM playerRepository) {
            super(plugin, playerRepository);
        }

        @PostConstruct
        public void init() {
            // disconnect any connected player...
            for (var player : ProxyServer.getInstance().getPlayers()) {
                player.disconnect();
            }

            getPlugin().getProxy().getPluginManager().registerListener(getPlugin(), this);
        }

        // should always be executed first.
        @net.md_5.bungee.event.EventHandler(priority = net.md_5.bungee.event.EventPriority.LOW)
        public void onPostLogin(PostLoginEvent e) {
            handlePlayerJoin(e.getPlayer().getUniqueId(), e.getPlayer());
        }

        // should always be executed last.
        @net.md_5.bungee.event.EventHandler(priority = net.md_5.bungee.event.EventPriority.HIGH)
        public void onPlayerDisconnect(PlayerDisconnectEvent e) {
            handlePlayerQuit(e.getPlayer().getUniqueId());
        }

        @Override
        protected @NonNull Class<ProxiedPlayer> playerType() {
            return ProxiedPlayer.class;
        }
    }
}

File: vital-players/src/main/kotlin/me/vitalframework/players/VitalPlayerListener.kt
package me.vitalframework.players

import me.vitalframework.BungeeEventHandler
import me.vitalframework.BungeeEventPriority
import me.vitalframework.BungeePlayer
import me.vitalframework.SpigotEventHandler
import me.vitalframework.SpigotEventPriority
import me.vitalframework.SpigotPlayer
import me.vitalframework.VitalListener
import net.md_5.bungee.api.event.PlayerDisconnectEvent
import net.md_5.bungee.api.event.PostLoginEvent
import org.bukkit.event.inventory.InventoryCloseEvent
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerQuitEvent
import java.util.UUID

/**
 * Internal listener; manages [VitalPlayer] lifecycles.
 */
interface VitalPlayerListener {
    val vitalPlayerService: VitalPlayerService
    val vitalPlayersConfigurationProperties: VitalPlayersConfigurationProperties

    /**
     * Internal function; creates a new [VitalPlayer] instance for the given [player] and stores it in [VitalPlayerRepository].
     * This function throws [VitalPlayerException.InvalidClass] if the configured custom player class has an invalid signature.
     */
    fun <T : Any> createPlayer(
        player: T,
        playerUniqueId: UUID,
        playerClass: Class<T>,
    ): VitalPlayer<*> =
        try {
            vitalPlayerService.createPlayer(
                player,
                playerUniqueId,
                playerClass,
                vitalPlayersConfigurationProperties.playerClass,
            )
        } catch (e: ClassCastException) {
            throw VitalPlayerException.InvalidClass(vitalPlayersConfigurationProperties.playerClass, e)
        }

    /**
     * Destroys the [VitalPlayer] instance of the given [playerUniqueId].
     * If no [VitalPlayer] instance exists for the given [playerUniqueId], this function does nothing.
     */
    fun destroyPlayer(playerUniqueId: UUID) = vitalPlayerService.deletePlayer(playerUniqueId)

    class Spigot(
        override val vitalPlayerService: VitalPlayerService,
        override val vitalPlayersConfigurationProperties: VitalPlayersConfigurationProperties,
    ) : VitalListener.Spigot(),
        VitalPlayerListener {
        // should always be executed first.
        @SpigotEventHandler(SpigotEventPriority.LOWEST)
        fun onPlayerJoin(e: PlayerJoinEvent) {
            createPlayer(e.player, e.player.uniqueId, SpigotPlayer::class.java)
        }

        // should always be executed last.
        @SpigotEventHandler(SpigotEventPriority.HIGHEST)
        fun onPlayerQuit(e: PlayerQuitEvent) {
            // before destroying the vital player instance, we must first close the active inventory.
            e.player.closeInventory(InventoryCloseEvent.Reason.DISCONNECT)
            destroyPlayer(e.player.uniqueId)
        }
    }

    class Bungee(
        override val vitalPlayerService: VitalPlayerService,
        override val vitalPlayersConfigurationProperties: VitalPlayersConfigurationProperties,
    ) : VitalListener.Bungee(),
        VitalPlayerListener {
        // should always be executed first.
        @BungeeEventHandler(BungeeEventPriority.LOWEST)
        fun onPostLogin(e: PostLoginEvent) {
            createPlayer(e.player, e.player.uniqueId, BungeePlayer::class.java)
        }

        // should always be executed last.
        @BungeeEventHandler(BungeeEventPriority.HIGHEST)
        fun onPlayerDisconnect(e: PlayerDisconnectEvent) {
            destroyPlayer(e.player.uniqueId)
        }
    }
}
